src/core/lib/gpr/tls.h

```diff
@@ -21,52 +21,121 @@
-     gpr_tls_set(&foo, new_value);
+template <typename T, typename = typename std::enable_if<(
+                          std::is_trivially_destructible<T>::value &&
+                          sizeof(T) <= sizeof(void*) &&
+                          alignof(void*) % alignof(T) == 0)>::type>
+class TlsTypeWrapper {
+  static_assert(std::is_trivially_destructible<T>::value &&
+                    sizeof(T) <= sizeof(void*) &&
+                    alignof(void*) % alignof(T) == 0,
+                "unsupported type for TLS");
 
-   Accessing a thread local:
-     current_value = gpr_tls_get(&foo);
+ public:
+  using Type = T;
+};
 
-   ALL functions here may be implemented as macros. */
+}  // namespace grpc_core
 
 #ifdef GPR_STDCPP_TLS
-#include "src/core/lib/gpr/tls_stdcpp.h"
+#define GPR_THREAD_LOCAL(type) \
+  thread_local grpc_core::TlsTypeWrapper<type>::Type
 #endif
 
 #ifdef GPR_GCC_TLS
-#include "src/core/lib/gpr/tls_gcc.h"
+#define GPR_THREAD_LOCAL(type) __thread grpc_core::TlsTypeWrapper<type>::Type
 #endif
 
 #ifdef GPR_MSVC_TLS
-#include "src/core/lib/gpr/tls_msvc.h"
+#define GPR_THREAD_LOCAL(type) \
+  __declspec(thread) grpc_core::TlsTypeWrapper<type>::Type
 #endif
 
 #ifdef GPR_PTHREAD_TLS
-#include "src/core/lib/gpr/tls_pthread.h"
+#include <grpc/support/log.h> /* for GPR_ASSERT */
+#include <pthread.h>
+
+namespace grpc_core {
+
+template <typename T>
+class pthreadTlsWrapper : TlsTypeWrapper<T> {
+ public:
+  pthreadTlsWrapper() = default;
+  pthreadTlsWrapper(const pthreadTlsWrapper&) = delete;
+  pthreadTlsWrapper& operator=(const pthreadTlsWrapper&) = delete;
+
+  void Init() { GPR_ASSERT(0 == pthread_key_create(&key_, nullptr)); }
+
+  void Destroy() { GPR_ASSERT(0 == pthread_key_delete(key_)); }
+
+  operator T() const { return Cast<T>(pthread_getspecific(key_)); }
+
+  T operator=(T t) {
+    GPR_ASSERT(0 == pthread_setspecific(key_, Cast<T>(t)));
+    return t;
+  }
+
+ private:
+  // TODO(C++17): Replace these helpers with constexpr if statements inline.
+  template <typename V>
+  static typename std::enable_if<std::is_pointer<V>::value, V>::type Cast(
+      void* object) {
+    return static_cast<V>(object);
+  }
+
+  template <typename V>
+  static typename std::enable_if<
+      !std::is_pointer<V>::value && std::is_integral<V>::value, V>::type
+  Cast(void* object) {
+    return reinterpret_cast<uintptr_t>(object);
+  }
+
+  template <typename V>
+  static void* Cast(
+      typename std::enable_if<std::is_pointer<V>::value, V>::type t) {
+    return t;
+  }
+
+  template <typename V>
+  static void* Cast(typename std::enable_if<!std::is_pointer<V>::value &&
+                                                std::is_integral<V>::value,
+                                            V>::type t) {
+    return reinterpret_cast<void*>(uintptr_t(t));
+  }
+
+  pthread_key_t key_;
+};
+
+}  // namespace grpc_core
+
+#define GPR_THREAD_LOCAL(type) grpc_core::pthreadTlsWrapper<type>
```

Comment:
It looks like there are a number of cases where we use a global thread-local variable.  Note that the style guide prohibits global variables that are not trivially destructible:

https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

Is the `pthreadTlsWrapper` class trivially destructible?  If not, this won't work.  If so, please add a static assertion that guarantees that property, with a comment as to why it's there.

Reply:
Done.